### PR TITLE
[Feature](network interface) Support network interface

### PR DESCRIPTION
**Pull request by ZashJie**
# Proposed changes
Support network_interface in Apache Doris, the main changes are:

1. If network_interface is configured in the configuration file, use network_interface configuration

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [x] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hf200012** [comment]
title style ：[<type>](<scope>) <subject> 
https://doris.apache.org/zh-CN/community/how-to-contribute/commit-format-specification
@ZashJie

File: be/src/service/backend_options.cpp [hunk 1]
```diff
@@ -81,15 +84,11 @@ std::string BackendOptions::get_localhost() {
...
 }
 
-bool BackendOptions::analyze_priority_cidrs() {
-    if (config::priority_networks == "") {
-        return true;
-    }
-    LOG(INFO) << "priority cidrs in conf: " << config::priority_networks;
-
+// 解析IP为cidr表达方式
+bool BackendOptions::analyze_cidrs(const std::string & ip) {
```
Comment (dataroaring) [inline be/src/service/backend_options.cpp:88]: There is no need to modify this function.

File: be/src/service/backend_options.cpp [hunk 2]
```diff
@@ -101,6 +100,61 @@ bool BackendOptions::analyze_priority_cidrs() {
...
+
+//从network_interface中获取IP
+bool BackendOptions::analyze_network_interface() {
+    if (config::network_interface[0] == "") {
+        // 为空，则去分析priority_networks配置值
+        return false;
+    }
+    LOG(INFO) << "network name in conf: " << config::network_interface[0];
+    LOG(INFO) << "priority cidrs in conf: " << config::network_interface[1];
+
```
Comment (dataroaring) [inline be/src/service/backend_options.cpp:112]: Please merge the above 2 LOGS to one, LOG(INFO) << "" << "" <<.

**Review by dataroaring — CHANGES_REQUESTED** [review]

File: be/src/service/backend_options.cpp [hunk 3]
```diff
@@ -81,15 +84,11 @@ std::string BackendOptions::get_localhost() {
     return _s_localhost;
 }
 
-bool BackendOptions::analyze_priority_cidrs() {
-    if (config::priority_networks == "") {
-        return true;
-    }
-    LOG(INFO) << "priority cidrs in conf: " << config::priority_networks;
-
+// 解析IP为cidr表达方式
```
Comment (Yukang-Lian) [inline be/src/service/backend_options.cpp:87]: please use english comment.

File: be/src/service/backend_options.cpp [hunk 4]
```diff
@@ -101,6 +100,61 @@ bool BackendOptions::analyze_priority_cidrs() {
     return true;
 }
 
+
+//从network_interface中获取IP
```
Comment (Yukang-Lian) [inline be/src/service/backend_options.cpp:104]: ditto

File: be/src/service/backend_options.cpp [hunk 5]
```diff
@@ -101,6 +100,61 @@ bool BackendOptions::analyze_priority_cidrs() {
...
+    
+    // for (auto& cidr_str : cidr_strs) {
+    //     CIDR cidr;
+    //     if (!cidr.reset(cidr_str)) {
+    //         LOG(FATAL) << "wrong cidr format. cidr_str=" << cidr_str;
+    //         return false;
+    //     }
+    //     _s_priority_cidrs.push_back(cidr);
+    // }
+    // return true;
```
Comment (Yukang-Lian) [inline be/src/service/backend_options.cpp:129]: useless code?

File: be/src/service/backend_options.cpp [hunk 6]
```diff
@@ -101,6 +100,61 @@ bool BackendOptions::analyze_priority_cidrs() {
...
+}
```
Comment (Yukang-Lian) [inline be/src/service/backend_options.cpp:156]: ditto

File: conf/be.conf [hunk 7]
```diff
@@ -34,13 +34,14 @@ brpc_port = 8060
...
+network_interface = eth1, 127.0.0.1/16
 
 # data root path, separate by ';'
 # you can specify the storage medium of each root path, HDD or SSD
 # you can add capacity limit at the end of each root path, separate by ','
 # eg:
-# storage_root_path = /home/disk1/doris.HDD,50;/home/disk2/doris.SSD,1;/home/disk2/doris
+# storage_root_path = /root/git/output/be/storage;/home/disk2/doris.SSD,1;/home/disk2/doris
 # /home/disk1/doris.HDD, capacity limit is 50GB, HDD;
 # /home/disk2/doris.SSD, capacity limit is 1GB, SSD;
```
Comment (Yukang-Lian) [inline conf/be.conf:46]: modify this file in your environment, but do not commit it.

File: conf/fe.conf [hunk 8]
```diff
@@ -52,7 +52,7 @@ mysql_service_nio_enabled = true
 # If no ip match this rule, will choose one randomly.
 # use CIDR format, e.g. 10.10.10.0/24
 # Default value is empty.
-# priority_networks = 10.10.10.0/24;192.168.0.0/16
+priority_networks = 127.0.0.1/16
 
 # Advanced configurations 
 # log_roll_size_mb = 1024
```
Comment (Yukang-Lian) [inline conf/fe.conf:58]: ditto

**Review by Yukang-Lian — CHANGES_REQUESTED** [review]
please change your PR title format.

**Comment by morningman** [comment]
Please amend your PR title and add describe in your PR comment.
And I suggest to wait #14063 merge

File: be/src/common/config.h [hunk 9]
```diff
@@ -49,6 +49,8 @@ CONF_Int32(single_replica_load_brpc_num_threads, "64");
 // If no ip match this rule, will choose one randomly.
 CONF_String(priority_networks, "");
 
+CONF_Strings(network_interface, "");
```
Comment (dataroaring) [inline be/src/common/config.h:52]: String is enough.

File: be/src/service/backend_options.cpp [hunk 10]
```diff
@@ -81,6 +84,51 @@ std::string BackendOptions::get_localhost() {
     return _s_localhost;
 }
 
+bool BackendOptions::analyze_network_interface() {
+    if (config::network_interface[0] == "") {
+        return false;
+    }
+    LOG(INFO) << "network name in conf: " << config::network_interface[0] << "\tpriority cidrs in conf: " << config::network_interface[1];
+
```
Comment (dataroaring) [inline be/src/service/backend_options.cpp:92]: we should not print cidrs here, it is confusing.

File: be/src/service/backend_options.cpp [hunk 11]
```diff
@@ -81,6 +84,51 @@ std::string BackendOptions::get_localhost() {
     return _s_localhost;
 }
 
+bool BackendOptions::analyze_network_interface() {
+    if (config::network_interface[0] == "") {
+        return false;
+    }
+    LOG(INFO) << "network name in conf: " << config::network_interface[0] << "\tpriority cidrs in conf: " << config::network_interface[1];
+    
+    std::vector<std::string> cidr_strs =
```
Comment (dataroaring) [inline be/src/service/backend_options.cpp:93]: variable name is confusing.

**Review by dataroaring — CHANGES_REQUESTED** [review]

File: be/src/service/backend_options.cpp [hunk 12]
```diff
@@ -57,6 +62,10 @@ bool BackendOptions::init() {
             VLOG_CRITICAL << "check ip=" << addr_it->get_host_address_v4();
             if ((*addr_it).is_loopback_v4()) {
                 loopback = addr_it->get_host_address_v4();
+            } else if (!_s_net_interfaces.empty()) {
+                if (is_in_net_interface(addr_it->get_host_address_v4())) {
+                    _s_localhost = addr_it->get_host_address_v4();
+                }
```
Comment (dataroaring) [inline be/src/service/backend_options.cpp:68]: We should not follow cidrs here.

There is a problem here, e.g. There are 2 nics, eth0 and eth1, we configed eth1 and cidrs, if we tranverse eth0 first then cidr is used. we should let network interface is selected in priority. 

So we should handle  net_interface before this while.

File: be/src/service/backend_options.h [hunk 13]
```diff
@@ -21,6 +21,8 @@
 
 #include <string>
 #include <vector>
+#include <ifaddrs.h>
+#include <arpa/inet.h>
```
Comment (dataroaring) [inline be/src/service/backend_options.h:25]: put the two .h to .cpp .

File: fe/fe-core/src/main/java/org/apache/doris/service/FrontendOptions.java [hunk 14]
```diff
@@ -68,6 +81,11 @@ public static void init() throws UnknownHostException {
             if (addr instanceof Inet4Address) {
                 if (addr.isLoopbackAddress()) {
                     loopBack = addr;
+                } else if (!netInterfaces.isEmpty()) {
+                    if (isInNetInterface(addr.getHostAddress())) {
+                        localAddr = addr;
+                        break;
+                    }
```
Comment (dataroaring) [inline fe/fe-core/src/main/java/org/apache/doris/service/FrontendOptions.java:88]: same as be.

**Review by dataroaring — CHANGES_REQUESTED** [review]

**Review by dataroaring — CHANGES_REQUESTED** [review]
Please have a look at the comments.

**Review by dataroaring — DISMISSED** [review]
LGTM

**Pull request closed without merging**